In [15]:
%matplotlib inline

import numpy as np
import matplotlib.pyplot as plt
from keras.datasets import mnist
from keras.utils import np_utils

In [16]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train = x_train.reshape(60000, 784)
x_test = x_test.reshape(10000, 784)
y_train = np_utils.to_categorical(y_train, 10)
y_test = np_utils.to_categorical(y_test, 10)

In [17]:
from keras.models import Sequential
from keras.layers import Dense,Activation
from keras.optimizers import SGD

## 第一個model:上課範例

In [18]:
model_1 = Sequential()
model_1.add(Dense(4, input_dim = 28*28))
model_1.add(Activation('softplus'))
model_1.add(Dense(2))
model_1.add(Activation('softplus'))
model_1.add(Dense(10))
model_1.add(Activation('softmax'))
model_1.summary()

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_19 (Dense)             (None, 4)                 3140      
_________________________________________________________________
activation_19 (Activation)   (None, 4)                 0         
_________________________________________________________________
dense_20 (Dense)             (None, 2)                 10        
_________________________________________________________________
activation_20 (Activation)   (None, 2)                 0         
_________________________________________________________________
dense_21 (Dense)             (None, 10)                30        
_________________________________________________________________
activation_21 (Activation)   (None, 10)                0         
Total params: 3,180
Trainable params: 3,180
Non-trainable params: 0
_________________________________________________________________


In [19]:
model_1.compile(loss = "mse",optimizer=SGD(lr = 0.087),metrics=["accuracy"]) 

In [20]:
model_1.fit(x_train,y_train, batch_size=100,epochs=20)

Epoch 1/20
60000/60000 [==============================] - 1s 11us/step - loss: 0.0865 - acc: 0.1030
Epoch 2/20
60000/60000 [==============================] - 0s 7us/step - loss: 0.0807 - acc: 0.2579
Epoch 3/20
60000/60000 [==============================] - 0s 7us/step - loss: 0.0796 - acc: 0.2847
Epoch 4/20
60000/60000 [==============================] - 0s 7us/step - loss: 0.0790 - acc: 0.2908
Epoch 5/20
60000/60000 [==============================] - 0s 7us/step - loss: 0.0782 - acc: 0.2954
Epoch 6/20
60000/60000 [==============================] - 0s 7us/step - loss: 0.0776 - acc: 0.2967
Epoch 7/20
60000/60000 [==============================] - 0s 7us/step - loss: 0.0770 - acc: 0.2980
Epoch 8/20
60000/60000 [==============================] - 0s 7us/step - loss: 0.0765 - acc: 0.2983
Epoch 9/20
60000/60000 [==============================] - 0s 7us/step - loss: 0.0760 - acc: 0.2989
Epoch 10/20
60000/60000 [==============================] - 0s 7us/step - loss: 0.0756 - acc: 0.2990
Epoch 11

正確率低，約30%

## 第二個model:調整learning rate

In [21]:
model_2 = Sequential()
model_2.add(Dense(10, input_dim = 28*28))
model_2.add(Activation('softplus'))
model_2.add(Dense(10))
model_2.add(Activation('softplus'))
model_2.add(Dense(10))
model_2.add(Activation('softmax'))
model_2.summary()
model_2.compile(loss = "mse",optimizer=SGD(lr = 0.1),metrics=["accuracy"]) 
model_2 = model_2.fit(x_train,y_train,batch_size=100,epochs=20)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_22 (Dense)             (None, 10)                7850      
_________________________________________________________________
activation_22 (Activation)   (None, 10)                0         
_________________________________________________________________
dense_23 (Dense)             (None, 10)                110       
_________________________________________________________________
activation_23 (Activation)   (None, 10)                0         
_________________________________________________________________
dense_24 (Dense)             (None, 10)                110       
_________________________________________________________________
activation_24 (Activation)   (None, 10)                0         
Total params: 8,070
Trainable params: 8,070
Non-trainable params: 0
_________________________________________________________________
Epoch 1/

調高learning rate，正確率上升：大約從30%上升到63%

## 第三個model:調整神經元數量

In [22]:
model_3= Sequential()
model_3.add(Dense(20, input_dim = 28*28))
model_3.add(Activation('hard_sigmoid'))
model_3.add(Dense(20))
model_3.add(Activation('hard_sigmoid'))
model_3.add(Dense(10))
model_3.add(Activation('hard_sigmoid'))
model_3.summary()
model_3.compile(loss = "mse",optimizer=SGD(lr = 0.087),metrics=["accuracy"]) #告知loss function
model_3 =model_3.fit(x_train,y_train,batch_size=100,epochs=20)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_25 (Dense)             (None, 20)                15700     
_________________________________________________________________
activation_25 (Activation)   (None, 20)                0         
_________________________________________________________________
dense_26 (Dense)             (None, 20)                420       
_________________________________________________________________
activation_26 (Activation)   (None, 20)                0         
_________________________________________________________________
dense_27 (Dense)             (None, 10)                210       
_________________________________________________________________
activation_27 (Activation)   (None, 10)                0         
Total params: 16,330
Trainable params: 16,330
Non-trainable params: 0
_________________________________________________________________
Epoch 

將第一二層的神經元都增加到20個，正確率大幅提升到約76%
上升程度比調高learning rate高一些

## 第四個模型：增加層數(三層)
因為增加神經元有效提高正確率，所以各層維持較多的神經元數量
因為提高learning rate有效提高正確率，所以維持較高的learning rate

In [26]:
model_4 = Sequential()
model_4.add(Dense(20, input_dim = 28*28))
model_4.add(Activation('softplus'))
model_4.add(Dense(20))
model_4.add(Activation('softplus'))
model_4.add(Dense(20))
model_4.add(Activation('softplus'))
model_4.add(Dense(10))
model_4.add(Activation('softmax'))
model_4.summary()
model_4.compile(loss = "mse",optimizer=SGD(lr = 0.1),metrics=["accuracy"]) 
model_4 =model_4.fit(x_train,y_train,batch_size=100,epochs=20)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_40 (Dense)             (None, 20)                15700     
_________________________________________________________________
activation_40 (Activation)   (None, 20)                0         
_________________________________________________________________
dense_41 (Dense)             (None, 20)                420       
_________________________________________________________________
activation_41 (Activation)   (None, 20)                0         
_________________________________________________________________
dense_42 (Dense)             (None, 20)                420       
_________________________________________________________________
activation_42 (Activation)   (None, 20)                0         
_________________________________________________________________
dense_43 (Dense)             (None, 10)                210       
__________

正確率提高至大約88%

## 第五個模型：調整batch_size & epoch

In [28]:
model_5 = Sequential()
model_5.add(Dense(20, input_dim = 28*28))
model_5.add(Activation('softplus'))
model_5.add(Dense(20))
model_5.add(Activation('softplus'))
model_5.add(Dense(20))
model_5.add(Activation('softplus'))
model_5.add(Dense(10))
model_5.add(Activation('softmax'))
model_5.summary()
model_5.compile(loss = "mse",optimizer=SGD(lr = 0.1),metrics=["accuracy"]) 
model_5 =model_5.fit(x_train,y_train,batch_size=100,epochs=100)

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_48 (Dense)             (None, 20)                15700     
_________________________________________________________________
activation_48 (Activation)   (None, 20)                0         
_________________________________________________________________
dense_49 (Dense)             (None, 20)                420       
_________________________________________________________________
activation_49 (Activation)   (None, 20)                0         
_________________________________________________________________
dense_50 (Dense)             (None, 20)                420       
_________________________________________________________________
activation_50 (Activation)   (None, 20)                0         
_________________________________________________________________
dense_51 (Dense)             (None, 10)                210       
__________

60000/60000 [==============================] - 1s 10us/step - loss: 0.0058 - acc: 0.9650
Epoch 68/100
60000/60000 [==============================] - 1s 10us/step - loss: 0.0058 - acc: 0.9650
Epoch 69/100
60000/60000 [==============================] - 1s 11us/step - loss: 0.0058 - acc: 0.9647
Epoch 70/100
60000/60000 [==============================] - 1s 10us/step - loss: 0.0058 - acc: 0.9651
Epoch 71/100
60000/60000 [==============================] - 1s 10us/step - loss: 0.0057 - acc: 0.9653
Epoch 72/100
60000/60000 [==============================] - 1s 11us/step - loss: 0.0057 - acc: 0.9655
Epoch 73/100
60000/60000 [==============================] - 1s 11us/step - loss: 0.0057 - acc: 0.9656
Epoch 74/100
60000/60000 [==============================] - 1s 10us/step - loss: 0.0055 - acc: 0.9667
Epoch 75/100
60000/60000 [==============================] - 1s 11us/step - loss: 0.0056 - acc: 0.9666
Epoch 76/100
60000/60000 [==============================] - 1s 11us/step - loss: 0.0056 - acc: 

調高epochs，正確率提升到約97%

# 結論

認為可能可以提高正確率的方法：

1.調高learning rate

2.增加每層神經元數量

3.增加層數

4.調高epochs